geo/geodist: speed up dwithin/distance operations with bbox

otan opened the pull request:
We can optimize distance checks by saying if their bounding boxes do not
intersect then we only need to check the exteriors. This shaved 2-3s off
a 7s query available in the PostGIS tutorial.

Existing test cases exercise this functionality well.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49799)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49799#-:-M8rXpAE6bsXsb_4g1XZ:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geodist/geodist.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49799#-M8rWSF_B2K88YyqmGrD:-M8rWSF_B2K88YyqmGrE:b-qt38pv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f72065a53a59a3bd19d64675a37c21f20fc1a45e/pkg/geo/geodist/geodist.go#L245)):*
> ```Go
> 			// BoundingBoxIntersects: if the bounding box of the two shapes do not intersect,
> 			//   then we don't need to check whether edges intersect either.
> 			if !c.DistanceUpdater().IsMaxDistance() && c.BoundingBoxIntersects() {
> ```

This comment could be hoisted a bit earlier in this function and cleanly separated into the MinDistance and MaxDistance cases. Only the former needs crosser if the bounding boxes intersect.
And then this if-condition could be simplified to
`if crosser != nil`
___
*[pkg/geo/geogfn/distance.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49799#-M8rOvYq2WstzXQalCaI:-M8rOvYr91CBt_k15hzm:b-geya6j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f72065a53a59a3bd19d64675a37c21f20fc1a45e/pkg/geo/geogfn/distance.go#L309)):*
> ```Go
> func (c *geographyDistanceCalculator) BoundingBoxIntersects() bool {
> 	// Return true, as it does the safer thing beneath.
> 	// TODO(otan): update bounding box intersects.
> ```

this is relying on knowing what the caller is doing with this, which seems error-prone. Since this is going to be short-lived this is fine for now, otherwise we should rename it to something like `BoundingBoxPossiblyIntersects`.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49799)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geodist/geodist.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49799#-M8rWSF_B2K88YyqmGrD:-M8r_3pB3bE2yCUdRF8P:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f72065a53a59a3bd19d64675a37c21f20fc1a45e/pkg/geo/geodist/geodist.go#L245)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This comment could be hoisted a bit earlier in this function and cleanly separated into the MinDistance and MaxDistance cases. Only the former needs crosser if the bounding boxes intersect.
And then this if-condition could be simplified to
`if crosser != nil`
</blockquote></details>

Done.
___
*[pkg/geo/geogfn/distance.go, line 309 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49799#-M8rOvYq2WstzXQalCaI:-M8rZsj7-IBhKMstqJMi:bnftpa4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f72065a53a59a3bd19d64675a37c21f20fc1a45e/pkg/geo/geogfn/distance.go#L309)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this is relying on knowing what the caller is doing with this, which seems error-prone. Since this is going to be short-lived this is fine for now, otherwise we should rename it to something like `BoundingBoxPossiblyIntersects`.
</blockquote></details>

It is temporary so I will remove this.


<!-- Sent from Reviewable.io -->

otan committed:
geo/geodist: speed up dwithin/distance operations with bbox

We can optimize distance checks by saying if their bounding boxes do not
intersect then we only need to check the exteriors. This shaved 2-3s off
a 7s query available in the PostGIS tutorial.

Existing test cases exercise this functionality well.

Release note: None

otan commented:
bors r=sumeerbhola
